### ISSUE TITLE
HTML sanitized when importing to Mongo

### ISSUE THREAD

**Issue by TomFrost** [akhoury/nodebb-plugin-import#112]
I've imported my IPB forum to both Redis and Mongo 3.0.x, and after the Redis import, html tags were intact and providing appropriate formatting.  In Mongo, however, my posts are littered with human-readable `<p></p>` and `<br />` and all the rest, in plain text -- they appear to have been sanitized on import.

Possibly related (but can be split into a new ticket if not), I noticed after my Mongo import that UTF-8 characters from IPB broke down into multiple ASCII characters after import. But I'm unfortunately unsure if this was also the case on Redis or not.

**Comment by akhoury** [comment]
Do you the "Markdown" plugin activated? if so, can you take a screenshot of its configuration page? I know it does sanitize HTML.

If that works, and you decide to disable HTML sanitization you'd still want to stay safe, use this plugin along with Markdown.

https://github.com/akhoury/nodebb-plugin-sanitizehtml

### another option

However, you can try to actually let the Import plugin convert your HTML to Markdown.
Admin Control Panel > Import > Post-Import Tools > Convert, select HTML to Markdown

The original content is preserved if you don't like what you get, so technically you can do the same step again, but this time set the option to "Don't touch my content" and it revert back to the original content.

**Comment by TomFrost** [comment]
I actually had the Markdown plugin removed for both during the initial import -- but I apologize; I misdiagnosed this issue!

My IPB forum has a mix of BBCode and HTML. I'm not sure if the current version mixes it or, since I've been running IPB for five years, maybe an old version did it and that data just stayed in the DB.  It seems like I can use the BBCode to Markdown converter -or- the HTML to Markdown converter.  Running one after the other seems to undo the changes the first one made; presumably because the changes are being made from a secondary copy of what was imported.

So it's not the case that Redis worked differently than Mongo; I'd apparently just run the post-processing scripts in the opposite order for Mongo than I did for Redis.

Is there any way to convert both HTML *and* BBCode?

**Comment by akhoury** [comment]
What you need is just the BBCode to Markdown only AND re-enable the Markdown plugin. 

The BBCode to Markdown conversion, under the hood, __actually converts the BBCode to HTML, then HTML to Markdown.__
https://github.com/akhoury/bbcode-to-markdown/blob/master/bbcode-to-markdown.js#L26-L28

Which version of the Import plugin are you using? I recently fixed a major bug with the BBCode converter? it's in `0.3.8`.

Some HTML tags are acceptable markdown by the way. You just might need to disable HTML-sanitization use that sanitizehtml plugin to stay safe.

whatever is not converting, can you provide me with some sample of the raw content so I can run the tests and debug them my self? Provide me one with the BBCode, one with HTML and one with BOTH if you them.

As for the other issue, can I ask you to file another ticket? with a basic input example and its output.

**Comment by TomFrost** [comment]
Ah! Thank you so much for the suggestion! I'm having some great success now.  The only issues I'm running into are stemming from IPB's odd mix of html and bbcode in the database, which may be a relic of me running this forum for years and IPB changing how it saves posts over time.  The preprocessing javascript feature has been a wonderful way to solve that, though-- I'm still tweaking this, but so far this has taken care of the crux of my issues (since most of them stem from URLs within html tags being processed as links to be converted to MD before being processed as HTML):

```javascript
content = content.replace(/<a[^>]+href=["']([^"']+)["'][^>]*>\1<\/a>/g, '$1');
content = content.replace(/<a[^>]+href=['"]([^'"]+)['"][^>]*>([^<]+)<\/a>/g, '[url="$1"]$2[/url]');
content = content.replace(/<img[^>]+src=['"]([^'"]+)['"][^>]*>/g, '[img]$1[/img]');
```

The only strangeness left is that it takes _two_ runs of the converter to hit all my posts, which is odd.  I set `content = "test"` in my pre-converter, and a small number of more recent posts didn't get replaced with "test" until I ran it a second time. I can debug that one myself, though.

Thanks for your help!
